### PR TITLE
Adding keyword recommendations after FileKeyword

### LINKED ISSUE #66319: Missing suggestions after `file` keyword
**Version Used**: 

VS 17.5 p2

**Steps to Reproduce**:

Type code using `file` keyword:

```csharp
//      ↓ here
file sealed class A { }
//   ~~~~~~
```

**Expected Behavior**:

Intellisense will suggest other keywords that will be used for declaration on types after `file` keyword, such as `class`, `struct`, `sealed`, `abstract`, `readonly`, etc..

**Actual Behavior**:

No suggestion.

image

image

image

> Intellisense behaves well if other keywords besides it, but `file` is an exception.

### PR DESCRIPTION
**Pull request by kinkeadian25**
Closes #66319

### CODE REVIEW

File: src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs [hunk 1]
```diff
@@ -256,6 +256,13 @@ public async Task TestAfterPublic()
 @"public $$");
         }
 
+        [Fact]
+        public async Task TestAfterFile()
+        {
+            await VerifyKeywordAsync(SourceCodeKind.Regular,
+@"file $$");
+        }
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs:264]: I'm thinking of a test case like this:

```csharp

file $$

class C
{
}
```

Does it work currently? Is it hard to support?

File: src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs [hunk 2]
```diff
@@ -9,111 +9,112 @@
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations
 {
+    [Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs:12]: Nice change!

Comment (kinkeadian25) [reply src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs:12]: Thank you! pretty happy with that one!

File: src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs (hunk 1 above)
Comment (kinkeadian25) [reply src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs:264]: added in new test and passed! Thanks Youssef!

File: src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs [hunk 3]
```diff
@@ -195,199 +196,203 @@ public async Task TestAfterNestedAttribute()
...
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact]
         public async Task TestAfterPublic()
             => await VerifyKeywordAsync(@"public $$");
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact]
+        public async Task TestAFterFile()
```
Comment (genlu) [inline src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs:254]:
```suggestion
        public async Task TestAfterFile()
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 4]
```diff
@@ -600,7 +600,7 @@ public static bool IsLocalFunctionDeclarationContext(this SyntaxTree syntaxTree,
                 if (container is IncompleteMemberSyntax incompleteMember)
                     return incompleteMember.Type.IsKind(SyntaxKind.RefType);
 
-                if (container is CompilationUnitSyntax or BaseNamespaceDeclarationSyntax or TypeDeclarationSyntax)
+                if (container is CompilationUnitSyntax or BaseNamespaceDeclarationSyntax or TypeDeclarationSyntax or VariableDeclarationSyntax)
```
Comment (genlu) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:603]: Why is this change needed?

Comment (kinkeadian25) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:603]: FileKeyword parses to VariableDeclarationSyntax, to avoid a false return had to add that in. I noticed it caused a fail with a struct test. I am working on that now.

**Review by Youssef1313 — APPROVED** [review]
LGTM

**Review by genlu — APPROVED** [review]
Thanks

**Comment by genlu** [comment]
Oh, could you please add `[WorkItem(66319, "https://github.com/dotnet/roslyn/issues/66319")]` to the new tests?

**Comment by kinkeadian25** [comment]
> Oh, could you please add `[WorkItem(66319, "https://github.com/dotnet/roslyn/issues/66319")]` to the new tests?

Yeah! On it now!

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 5]
```diff
@@ -602,6 +602,9 @@ public static bool IsLocalFunctionDeclarationContext(this SyntaxTree syntaxTree,
 
                 if (container is CompilationUnitSyntax or BaseNamespaceDeclarationSyntax or TypeDeclarationSyntax)
                     return true;
+
+                if (container is VariableDeclarationSyntax && modifierTokens.Contains(SyntaxKind.FileKeyword))
+                    return true;
```
Comment (RikkiGibson) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:607]: What's the scenario where we hit this code path?

Comment (kinkeadian25) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:607]: That is what catches anything that my tests cover for that method. I made it independent, because when I added VariableDeclarationSyntax to the above if statement it caused some conflicts with other tests and recommenders. So, I have that there on its own to make sure that VariableDeclarationSyntax is only caught if one of the modifiers is FIleKeyword.

**Review by Youssef1313 — COMMENTED** [review]
This will need tests for

```csharp
namespace S
{
    file $$
}
```

and also

```csharp
namespace S
{
    file $$
    
    class C { }
}
```

File: src/EditorFeatures/CSharpTest2/Recommendations/ClassKeywordRecommenderTests.cs [hunk 6]
```diff
@@ -280,6 +280,13 @@ public async Task TestAfterPublic()
 @"public $$");
         }
 
+        [Fact, WorkItem(66319, "https://github.com/dotnet/roslyn/issues/66319")]
+        public async Task TestAfterFile()
+        {
+            await VerifyKeywordAsync(SourceCodeKind.Regular,
+@"file $$");
+        }
+
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest2/Recommendations/ClassKeywordRecommenderTests.cs:289]: Needs a test for record recommender as well.

Comment (kinkeadian25) [reply src/EditorFeatures/CSharpTest2/Recommendations/ClassKeywordRecommenderTests.cs:289]: Sounds good!

**Review by Youssef1313 — APPROVED** [review]
LGTM, assuming all tests pass :)

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Merge at will @genlu
